### PR TITLE
[SPARK-49709][CONNECT][SQL] Support ConfigEntry in the RuntimeConfig interface

### PR DESCRIPTION
**Pull request by hvanhovell**
### What changes were proposed in this pull request?
This PR adds support for ConfigEntry to the RuntimeConfig interface. This was removed in https://github.com/apache/spark/pull/47980.

### Why are the changes needed?
This functionality is used a lot by Spark libraries. Removing them caused friction, and adding them does not pollute the RuntimeConfig interface.

### Does this PR introduce _any_ user-facing change?
No. This is developer API.

### How was this patch tested?
I have added tests cases for Connect and Classic.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Review by dongjoon-hyun — APPROVED** [review]
Thank you for polishing this 😄. I'm wondering what makes you change your mind. As you know, I raised some issues from the original PR, @hvanhovell . Anyway, +1 for this direction.

> Removing them caused friction

File: sql/core/src/test/scala/org/apache/spark/sql/RuntimeConfigSuite.scala [hunk 1]
```diff
@@ -108,4 +108,26 @@ class RuntimeConfigSuite extends SparkFunSuite {
     // this set should not fail
     conf.set(DEFAULT_PARALLELISM.key, "1")
   }
+
+  test("config entry") {
+    val conf = newConf()
+
+    val entry = SQLConf.FILES_MAX_PARTITION_NUM
+    assert(conf.get(entry.key) === null)
```
Comment (hvanhovell) [inline sql/core/src/test/scala/org/apache/spark/sql/RuntimeConfigSuite.scala:116]: @HyukjinKwon @cloud-fan this returns null. I was expecting a `NoSuchElementException` here. IMO this is a bug. WDYT?

**Comment by hvanhovell** [comment]
Merging to master.

File: sql/core/src/test/scala/org/apache/spark/sql/RuntimeConfigSuite.scala (hunk 1 above)
Comment (HyukjinKwon) [reply sql/core/src/test/scala/org/apache/spark/sql/RuntimeConfigSuite.scala:116]: yeah, i think it should. `def get[T](entry: ConfigEntry[T]): T` isn't an API so I think we should be fine with changing this behaviour too.

**Pull request closed without merging**